toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1388 @@
+    // one when watching a resource.
+    if (mOnError) {
+      (void)mOnError->Complete(mError, mOsError);
+    }
+
+    return NS_OK;
+  }
+
+private:
+  nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback> mOnError;
+  nsresult mError;
+  long mOsError;
+};
+
+/**
+ * An event used to notify the main thread when an operation is successful.
+ */
+class WatchedSuccessEvent final : public Runnable
+{
+public:
+  /**
+   * @param aOnSuccess The passed success callback.
+   * @param aResourcePath
+   *        The path of the resource for which this event was generated.
+   */
+  WatchedSuccessEvent(const nsMainThreadPtrHandle<
+                        nsINativeFileWatcherSuccessCallback>& aOnSuccess,
+                      const nsAString& aResourcePath)
+    : Runnable("WatchedSuccessEvent")
+    , mOnSuccess(aOnSuccess)
+    , mResourcePath(aResourcePath)
+  {
+    MOZ_ASSERT(!NS_IsMainThread());
+  }
+
+  NS_IMETHOD Run() override
+  {
+    MOZ_ASSERT(NS_IsMainThread());
+
+    // Make sure we wrap a valid callback since it's not mandatory to provide
+    // one when watching a resource.
+    if (mOnSuccess) {
+      (void)mOnSuccess->Complete(mResourcePath);
+    }
+
+    return NS_OK;
+  }
+
+private:
+  nsMainThreadPtrHandle<nsINativeFileWatcherSuccessCallback> mOnSuccess;
+  nsString mResourcePath;
+};
+
+/**
+ * An event used to notify the main thread of a change in a watched
+ * resource.
+ */
+class WatchedChangeEvent final : public Runnable
+{
+public:
+  /**
+   * @param aOnChange The passed change callback.
+   * @param aChangedResource The name of the changed resource.
+   */
+  WatchedChangeEvent(
+    const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+    const nsAString& aChangedResource)
+    : Runnable("WatchedChangeEvent")
+    , mOnChange(aOnChange)
+    , mChangedResource(aChangedResource)
+  {
+    MOZ_ASSERT(!NS_IsMainThread());
+  }
+
+  NS_IMETHOD Run() override
+  {
+    MOZ_ASSERT(NS_IsMainThread());
+
+    // The second parameter is reserved for future uses: we use 0 as a
+    // placeholder.
+    (void)mOnChange->Changed(mChangedResource, 0);
+    return NS_OK;
+  }
+
+private:
+  nsMainThreadPtrHandle<nsINativeFileWatcherCallback> mOnChange;
+  nsString mChangedResource;
+};
+
+// Define these callback array types to make the code easier to read.
+typedef nsTArray<nsMainThreadPtrHandle<nsINativeFileWatcherCallback>>
+  ChangeCallbackArray;
+typedef nsTArray<nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>>
+  ErrorCallbackArray;
+
+static mozilla::LazyLogModule gNativeWatcherPRLog("NativeFileWatcherService");
+#define FILEWATCHERLOG(...)                                                    \
+  MOZ_LOG(gNativeWatcherPRLog, mozilla::LogLevel::Debug, (__VA_ARGS__))
+
+struct WatchedResourceDescriptor
+{
+  // The path on the file system of the watched resource.
+  nsString mPath;
+
+  // The watched descriptor returned from inotify_add_watch().
+  int mWatchedResourceDescriptor;
```

Comment:
In NativeFileWatcherWin.cpp, the destructor of `WatchedResourceDescriptor` closes this handle. You probably need to do the same here – if you don't, you definitely need to document how and where the resource needs to be closed.